### PR TITLE
Offer source writing and better connection handling

### PR DESCRIPTION
**Pull request by chamini2**

### CODE REVIEW

**Comment by chamini2** [comment]
My test script
```py
import pandas as pd
print(context)
model = context.current_model
print(ref(model.name))
data: pd.DataFrame = source('iris', 'data')
print(data)
write_to_source(data, 'iris', 'result')
print(source('iris', 'result'))
```

and the result:

```py
Context(meta={}, current_model=CurrentModel(name='null_class', status='success', columns={}))
   sepal_length sepal_width petal_length petal_width class
0           4.8         3.0          1.4         0.1  None
1           5.2         4.1          1.5         0.1  None
2           4.4         3.0          1.3         0.2  None
3           5.3         3.7          1.5         0.2  None
4           6.5         2.8          4.6         1.5  None
5           4.9         2.4          3.3         1.0  None
6           5.6         2.5          3.9         1.1  None
7           6.0         3.4          4.5         1.6  None
8           5.1         2.5          3.0         1.1  None
9           6.5         3.2          5.1         2.0  None
10          5.8         2.8          5.1         2.4  None
11          6.5         3.0          5.5         1.8  None
12          7.7         2.8          6.7         2.0  None
13          6.4         2.8          5.6         2.1  None
14          6.1         2.6          5.6         1.4  None
15          6.0         3.0          4.8         1.8  None
    sepal_length sepal_width petal_length petal_width           class
0            5.1         3.5          1.4         0.2     Iris-setosa
1            4.9         3.0          1.4         0.2     Iris-setosa
2            4.7         3.2          1.3         0.2     Iris-setosa
3            4.6         3.1          1.5         0.2     Iris-setosa
4            5.0         3.6          1.4         0.2     Iris-setosa
..           ...         ...          ...         ...             ...
145          6.7         3.0          5.2         2.3  Iris-virginica
146          6.3         2.5          5.0         1.9  Iris-virginica
147          6.5         3.0          5.2         2.0  Iris-virginica
148          6.2         3.4          5.4         2.3  Iris-virginica
149          5.9         3.0          5.1         1.8  Iris-virginica

[150 rows x 5 columns]
     sepal_length sepal_width petal_length petal_width           class
0             5.1         3.5          1.4         0.2     Iris-setosa
1             4.9         3.0          1.4         0.2     Iris-setosa
2             4.7         3.2          1.3         0.2     Iris-setosa
3             4.6         3.1          1.5         0.2     Iris-setosa
4             5.0         3.6          1.4         0.2     Iris-setosa
...           ...         ...          ...         ...             ...
3295          6.7         3.0          5.2         2.3  Iris-virginica
3296          6.3         2.5          5.0         1.9  Iris-virginica
3297          6.5         3.0          5.2         2.0  Iris-virginica
3298          6.2         3.4          5.4         2.3  Iris-virginica
3299          5.9         3.0          5.1         1.8  Iris-virginica

[3300 rows x 5 columns]
Context(meta={}, current_model=CurrentModel(name='iris_stats', status='success', columns={}))
  class count    avg_sepal_length     avg_sepal_width    avg_petal_length      avg_petal_width
0     C    45   5.971111111111109  2.7755555555555556   4.304444444444443   1.3333333333333333
1     G    43  6.6139534883720925   2.988372093023256   5.562790697674418   2.0418604651162786
2     S    46   5.013043478260869   3.415217391304348  1.4673913043478262  0.25217391304347825
    sepal_length sepal_width petal_length petal_width           class
0            5.1         3.5          1.4         0.2     Iris-setosa
1            4.9         3.0          1.4         0.2     Iris-setosa
2            4.7         3.2          1.3         0.2     Iris-setosa
3            4.6         3.1          1.5         0.2     Iris-setosa
4            5.0         3.6          1.4         0.2     Iris-setosa
..           ...         ...          ...         ...             ...
145          6.7         3.0          5.2         2.3  Iris-virginica
146          6.3         2.5          5.0         1.9  Iris-virginica
147          6.5         3.0          5.2         2.0  Iris-virginica
148          6.2         3.4          5.4         2.3  Iris-virginica
149          5.9         3.0          5.1         1.8  Iris-virginica

[150 rows x 5 columns]
     sepal_length sepal_width petal_length petal_width           class
0             5.1         3.5          1.4         0.2     Iris-setosa
1             4.9         3.0          1.4         0.2     Iris-setosa
2             4.7         3.2          1.3         0.2     Iris-setosa
3             4.6         3.1          1.5         0.2     Iris-setosa
4             5.0         3.6          1.4         0.2     Iris-setosa
...           ...         ...          ...         ...             ...
3445          6.7         3.0          5.2         2.3  Iris-virginica
3446          6.3         2.5          5.0         1.9  Iris-virginica
3447          6.5         3.0          5.2         2.0  Iris-virginica
3448          6.2         3.4          5.4         2.3  Iris-virginica
3449          5.9         3.0          5.1         1.8  Iris-virginica

[3450 rows x 5 columns]
Context(meta={}, current_model=CurrentModel(name='iris', status='success', columns={'sepal_length': ColumnInfo(name='sepal_length', description='', meta={}, data_type=None, quote=None, tags=[], _extra={'type': 'float'}), 'sepal_width': ColumnInfo(name='sepal_width', description='', meta={}, data_type=None, quote=None, tags=[], _extra={'type': 'float'}), 'petal_length': ColumnInfo(name='petal_length', description='', meta={}, data_type=None, quote=None, tags=[], _extra={'type': 'float'}), 'petal_width': ColumnInfo(name='petal_width', description='', meta={}, data_type=None, quote=None, tags=[], _extra={'type': 'float'}), 'class': ColumnInfo(name='class', description='', meta={}, data_type=None, quote=None, tags=[], _extra={'type': 'text'})}))
    sepal_length sepal_width petal_length petal_width class
0            5.1         3.5          1.4         0.2     S
1            4.9         3.0          1.4         0.2     S
2            4.7         3.2          1.3         0.2     S
3            4.6         3.1          1.5         0.2     S
4            5.0         3.6          1.4         0.2     S
..           ...         ...          ...         ...   ...
145          6.7         3.0          5.2         2.3     G
146          6.3         2.5          5.0         1.9     G
147          6.5         3.0          5.2         2.0     G
148          6.2         3.4          5.4         2.3     G
149          5.9         3.0          5.1         1.8     G

[150 rows x 5 columns]
    sepal_length sepal_width petal_length petal_width           class
0            5.1         3.5          1.4         0.2     Iris-setosa
1            4.9         3.0          1.4         0.2     Iris-setosa
2            4.7         3.2          1.3         0.2     Iris-setosa
3            4.6         3.1          1.5         0.2     Iris-setosa
4            5.0         3.6          1.4         0.2     Iris-setosa
..           ...         ...          ...         ...             ...
145          6.7         3.0          5.2         2.3  Iris-virginica
146          6.3         2.5          5.0         1.9  Iris-virginica
147          6.5         3.0          5.2         2.0  Iris-virginica
148          6.2         3.4          5.4         2.3  Iris-virginica
149          5.9         3.0          5.1         1.8  Iris-virginica

[150 rows x 5 columns]
     sepal_length sepal_width petal_length petal_width           class
0             5.1         3.5          1.4         0.2     Iris-setosa
1             4.9         3.0          1.4         0.2     Iris-setosa
2             4.7         3.2          1.3         0.2     Iris-setosa
3             4.6         3.1          1.5         0.2     Iris-setosa
4             5.0         3.6          1.4         0.2     Iris-setosa
...           ...         ...          ...         ...             ...
3595          6.7         3.0          5.2         2.3  Iris-virginica
3596          6.3         2.5          5.0         1.9  Iris-virginica
3597          6.5         3.0          5.2         2.0  Iris-virginica
3598          6.2         3.4          5.4         2.3  Iris-virginica
3599          5.9         3.0          5.1         1.8  Iris-virginica

[3600 rows x 5 columns]
```

File: src/fal/cli.py [hunk 1]
```diff
@@ -31,10 +34,21 @@
     is_flag=True,
     help="To only run models that ran in the last dbt run",
 )
-def run(run, project_dir, profiles_dir, keyword, all):
-    real_project_dir = os.path.realpath(os.path.normpath(project_dir))
-    real_profiles_dir = os.path.realpath(os.path.normpath(profiles_dir))
+@click.option(
+    "--debug",
+    is_flag=True,
+    help="To only run models that ran in the last dbt run",
```
Comment (turbo1912) [inline src/fal/cli.py:40]: wrong help

File: src/faldbt/lib.py [hunk 2]
```diff
@@ -45,60 +64,122 @@ def _get_operation_node(manifest: Manifest, project_path, sql):
...
+        agate_table, table_name=target.identifier, db_schema=target.schema
+    )
+
+    if relation is None:
+        stmt = CreateTable(alchemy_table).compile(
+            compile_kwargs={"literal_binds": True}
+        )
+        stmt_str = six.text_type(stmt).strip()
+
+        _execute_sql(manifest, project_path, stmt_str)
```
Comment (turbo1912) [inline src/faldbt/lib.py:169]: looks like these two lines are repeated below and can be after the if statement

File: src/faldbt/lib.py [hunk 3]
```diff
@@ -45,60 +64,122 @@ def _get_operation_node(manifest: Manifest, project_path, sql):
...
-def fetch_model(manifest: Manifest, project_path: str, model: ParsedModelNode):
-    adapter = _get_adapter(project_path)
-    relation = adapter.get_relation(model.database, model.schema, model.identifier)
-    query = f'SELECT * FROM {relation}'
-    return _execute_sql(manifest, project_path, adapter, query)
+def fetch_target(
+    manifest: Manifest,
+    project_path: str,
+    target: Union[ParsedModelNode, ParsedSourceDefinition],
+):
```
Comment (turbo1912) [inline src/faldbt/lib.py:132]: can we also add return types to these methods easier to follow

File: src/faldbt/lib.py [hunk 4]
```diff
@@ -45,60 +64,122 @@ def _get_operation_node(manifest: Manifest, project_path, sql):
...
+        .compile(compile_kwargs={"literal_binds": True})
+    )
+    stmt_str = six.text_type(stmt).strip()
+
+    _, result = _execute_sql(manifest, project_path, stmt_str)
+    return result
 
 
 def parse_to_manifest(config):
     from dbt.parser.manifest import ManifestLoader
```
Comment (turbo1912) [inline src/faldbt/lib.py:183]: is this import here for performance reasons?

File: src/faldbt/lib.py [hunk 5]
```diff
@@ -1,36 +1,55 @@
...
+from dbt.adapters.factory import Adapter
+from dbt.adapters.protocol import ConnectionManagerProtocol
+from dbt.config.runtime import RuntimeConfig
+from dbt.contracts.graph.manifest import Manifest
+import dbt.clients.agate_helper as agate_helper
+import dbt.adapters.factory as adapters_factory
+
+import pandas as pd
+
+from sqlalchemy.sql.ddl import CreateTable
```
Comment (turbo1912) [inline src/faldbt/lib.py:19]: looks like there are different dialects for each different source 

https://github.com/googleapis/python-bigquery-sqlalchemy

**Review by turbo1912 — APPROVED** [review]

File: src/fal/cli.py (hunk 1 above)
Comment (chamini2) [reply src/fal/cli.py:40]: thanks

File: src/faldbt/lib.py (hunk 4 above)
Comment (chamini2) [reply src/faldbt/lib.py:183]: It's how they did it, just never changed it. We can take it out. Do you prefer that?

File: src/faldbt/lib.py (hunk 5 above)
Comment (chamini2) [reply src/faldbt/lib.py:19]: Yes, I believe we generate for sqlite for now. Should we "match" the correct dialect in this PR or as a future task?

File: src/faldbt/lib.py (hunk 2 above)
Comment (chamini2) [reply src/faldbt/lib.py:169]: It is a different call, `stmt_str` is overwritten. Maybe I will change it.

File: src/faldbt/lib.py (hunk 4 above)
Comment (turbo1912) [reply src/faldbt/lib.py:182]: not really

**Pull request merged**